syntax = "proto3";

package envoy.config.core.v3;

// HeaderValue represents a single header key-value pair
message HeaderValue {
  string key = 1;
  string value = 2;
  bytes raw_value = 3;
}

// HeaderValueOption wraps HeaderValue with additional options
message HeaderValueOption {
  HeaderValue header = 1;
  // Deprecated: use append_action
  bool append = 2 [deprecated = true];
  enum HeaderAppendAction {
    APPEND_IF_EXISTS_OR_ADD = 0;
    ADD_IF_ABSENT = 1;
    OVERWRITE_IF_EXISTS_OR_ADD = 2;
    OVERWRITE_IF_EXISTS = 3;
  }
  HeaderAppendAction append_action = 3;
  bool keep_empty_value = 4;
}

// HeaderMap is a collection of headers
message HeaderMap {
  repeated HeaderValue headers = 1;
}

